src/app/server/Dnssd.cpp

```diff
@@ -214,7 +214,14 @@ CHIP_ERROR DnssdServer::AdvertiseOperational()
 #endif
 
 #if INET_CONFIG_ENABLE_TCP_ENDPOINT
-        advertiseParameters.SetTCPSupportModes(chip::Dnssd::TCPModeAdvertise::kTCPClientServer);
+        if (mTCPServerEnabled)
+        {
+            advertiseParameters.SetTCPSupportModes(chip::Dnssd::TCPModeAdvertise::kTCPClientServer);
+        }
+        else
+        {
+            advertiseParameters.SetTCPSupportModes(chip::Dnssd::TCPModeAdvertise::kTCPClient);
+        }
```

Comment:
I wonder if this would be more concise:

```suggestion
        advertiseParameters.SetTCPSupportModes(
            mTCPServerEnabled ? chip::Dnssd::TCPModeAdvertise::kTCPClientServer)
                              : chip::Dnssd::TCPModeAdvertise::kTCPClient);
```

Highlights that a TCP-supported-modes is always set and on fewer lines. Obviously subjective, so feel free to close if you diagree (this is not a PR blocker)